lib/pure/random.nim

```diff
@@ -617,15 +622,60 @@ proc shuffle*[T](x: var openArray[T]) =
 
   shuffle(state, x)
 
-when not defined(nimscript) and not defined(standalone):
-  import times
+when not defined(standalone):
+  when defined(nimscript):
+    import std/hashes
+
+    var baseState = block:
+      var ret = Rand(
+        a0: CompileTime.hash.Ui,
+        a1: CompileDate.hash.Ui)
+      if not ret.isValid:
+        ret = DefaultRandSeed
+      ret
+  elif defined(js):
+    import std/times
+  else:
+    import std/[hashes, os, sysrand, monotimes]
+
+    var baseState: Rand = block:
+      var
+        ret: Rand
+        urand: array[sizeof(Rand), byte]
+
+      for i in 0 .. 7:
+        if sysrand.urandom(urand):
+          copyMem(ret.addr, urand[0].addr, sizeof(Rand))
+          if ret.isValid:
+            break
+
+      if not ret.isValid:
+        # When 2 processes executed at same time on different machines,
+        # `ret` can still have same value.
+        let
+          pid = getCurrentProcessId()
+          t = getMonoTime().ticks
+        ret.a0 = pid.hash().Ui
+        ret.a1 = t.hash().Ui
+        if not ret.isValid:
+          ret.a0 = pid.Ui
+          ret.a1 = t.Ui
+
+      if not ret.isValid:
+        ret = DefaultRandSeed
+      ret
+
+    when compileOption("threads"):
```

Comment:
`baseState` cannot be a per-thread global because if it was per-thread, it cannot avoid random number sequences overlapping between different PRNGs in different thread.

Quote from https://prng.di.unimi.it (web site created by creator of xoroshiro):
> We provide ready-made jump functions for a number of calls equal to the square root of the period, to make it easy generating non-overlapping sequences for parallel computations, and equal to the cube of the fourth root of the period, to make it possible to generate independent sequences on different parallel processors.

`skipRandomNumbers*(s: var Rand)` in random module is a jump function.

Reply:
I don't follow - if per-thread state is initialized with random data via urandom, why would `skipRandomNumbers` overlap?

Reply:
Please read this:
https://www.mathcha.io/editor/7JPgyt0WSLdH4L7VywudZkBrQF4qJOe8Fp3JBB

Initial state of all `Rand` must be T2^64n (n is natural number including 0) to avoid overlapping.
If `baseState` were per-thread variable and they were initialized with `urandom`, initial state of all `Rand` cannot become T2^64n.

Reply:
@demotomohiro ...so now every call to next needs to hold the lock! I imagine this would be very slow. Is there a better solution than ignoring the data races?

Reply:
So if I understand correctly, you're stating that the problem is that `Rand` only has a single set of states (T0 to T2^64n), and that the initial seed only determines which of the 2^64n is the initial state of a particular `Rand` instance?

Reply:
@demotomohiro I'm also concerned about the point regarding performance cause by required locking raised in https://github.com/nim-lang/Nim/pull/18744#discussion_r701046207 ; could you please run some benchmarks to measure this?

note that the approach i proposed in https://github.com/nim-lang/Nim/pull/18149 did not have such problem

Reply:
@timotheecour The approach in https://github.com/nim-lang/Nim/pull/18149 would suffer from overlap issues though, wouldn't it? Which is what I believe @demotomohiro is aiming to avoid.

Reply:
> The approach in #18149 would suffer from overlap issues though, wouldn't it?

I was also ensuring `skipRandomNumbers` gets called but indeed, the way it was being called wouldn't guarantee lack of overlap ; i'll need to think more to double check;

that said, the potential performance issue should be evaluated/investigated

Reply:
Actually, upon re-reviewing the code:
  - The lock is only for initialization of new random states (from`basestate`), including the global state. I had previously misread this, and thought all access to the global random state (`state`) was being locked.
  - The current global random state isn't threadsafe at all (unless the implementation is a lock-free algorithm of some sort), and isn't a per-thread global. An obvious way to solve this to make `state` a thread-local global, and check/initialize it when it is accessed.

Reply:
@planetis-m
As @Varriount said, only `initRand()` proc acquires lock.
If you create `Rand` in each threads and never share them between threads, you don't need to lock when you call `next(r: var Rand)`.
If you create only one `Rand` and want to share it by multiple threads, you would need to acquire lock when you call `next(r: var Rand)`,
but it is not a problem this PR cause.

@Varriount 
In this document, T0 is a initial state of `baseState` and it is the value came from `sysrand.urandom`. 
https://www.mathcha.io/editor/7JPgyt0WSLdH4L7VywudZkBrQF4qJOe8Fp3JBB
Then, first `Rand` returned from `initRand()` has state T0.
second one has T2^64, third one has T2^64*2, 4th one has T2^64*3, n-th one has T2^64*(n-1) and so on.
Then, first `Rand` returned from `initRand()` can use states from T0 to T2^64-1 without overlapping the state with second one.
Second one can use from T2^64 to T2^64*2-1, third one can use from T2^64*2 to T2^64*3-1, and so on.

@timotheecour
I don't think performance of `initRand()` is so important, because it would not be called as frequently as `next(r: var Rand)` as `initRand()` is called only when you create a new `Rand`.
When many `Rand` are created but they generate few random numbers, performance of `initRand()` can be important.
But in that case, I think hash functions tested in paper would be better:
http://jcgt.org/published/0009/03/02/

> An obvious way to solve this to make `state` a thread-local global, and check/initialize it when it is accessed.

I think making `state` a thread-local variable and Nim compiler automatically inserts a call to `randomize()` so that it is called when a new thread starts would be better.
But I don't know it is possible.

Reply:
I get it now, good one +1